MSC3879: Trusted key forwards (#3879)
matrix-org/matrix-doc: proposals/3879-trusted-key-forwards.md

# MSC3879: Trusted key forwards

When a device does not have a megolm session, it can [request the session from
another device](https://spec.matrix.org/unstable/client-server-api/#key-requests).
However, when receiving sessions in this way, the device currently does not
know whether the session should be trusted.  As a result, some clients indicate
that the events encrypted with this session cannot be trusted, which causes
confusion for users.

We propose adding a flag to the `m.forwarded_room_key` event to indicate
whether the forwarding device believes the session should be trusted.  If the
recipient then trusts the forwarder, it can use that flag to decide whether or
not to trust the session.

## Proposal

A new property, `trusted`, is added to the `m.forwarded_room_key` event.  The
property is set to `true` if the forwarder believes the session should be
trusted.  If the property is set to `false` or is absent, this indicates that
the session cannot be trusted.

> The recipient can trust the session if:

Reviewer: ```suggestion
The recipient should trust the session iff all of the following is true:
```

Let's use a bit more normative language ("should" rather than "can") since we believe these conditions to be correct. Also "iff" or "if and only if" (should the former be too obscure).

- the `m.forwarded_room_key` event marks the key as trusted,

> - the recipient and forwarder belong to the same user, and

Reviewer: ```suggestion
- the recipient and forwarding devices belong to the same user, and
```

I think this is the intended meaning and I find it clearer to be explicit here.

- the forwarder has been verified (e.g. by verifying the device directly, or
  via cross-signing).

Note that this depends on the forwarder being able to determine whether it

> trusts the session.  The forwarder can mark the session as trusted if:

Reviewer: ```suggestion
trusts the session.  The forwarder can mark the session as trusted if any of
the following is true:
```

- it created the session;
- it received the session from the session creator via an `m.room_key` message;
- it was received via an `m.forwarded_room_key` from a trusted device belonging
  to the same user, and it was marked as trusted; or
- it was received from a key backup that can be trusted (such as [Symmetric key
  backup](https://github.com/matrix-org/matrix-spec-proposals/pull/3270)), and

>   was marked as trusted.

Reviewer: ```suggestion
  was marked as trusted in the backup.
```

> # MSC3879: Trusted key forwards

Author: As mentioned elsewhere, the term "trusted" isn't a great term to use.  It is used to mean various things in various places.  We'll probably switch to another term.

The basic premise of this MSC is to provide a method for the key forwarder to indicate whether it believes that the room key was actually sent by the given sender_key.  It is still up to the recipient of the key forward to determine whether the sender_key is "trusted" (e.g. by checking whether they have been verified)

Reviewer 2: > As mentioned elsewhere, the term "trusted" isn't a great term to use. It is used to mean various things in various places. We'll probably switch to another term.

My objection to "trusted" isn't particularly about the term itself, but rather that it tends to get used without any additional context (what does it mean for a device to be "trusted"?). Provided it's clear from the context what we actually mean by "trusted", it's fine.

Author: I think that "authenticated" would be a better term than "trusted".  I used "authenticated" for MSC4048, and I think we should use it here too.  In either case, both should use the same name.

When a device does not have a megolm session, it can [request the session from
another device](https://spec.matrix.org/unstable/client-server-api/#key-requests).
However, when receiving sessions in this way, the device currently does not
know whether the session should be trusted.  As a result, some clients indicate
that the events encrypted with this session cannot be trusted, which causes
confusion for users.

> We propose adding a flag to the `m.forwarded_room_key` event to indicate

Reviewer 3: Maybe clarify that it's added to the event content
